File: bench/benchmarks/binary_trees/binary_trees.js
const MAX_DEPTH = 14;

function treeSize(depth) {
    let size = 1;
    for (let i = 0; i <= depth; i++) size *= 2;
    return size - 1;
}

function buildAndCheck(depth) {
    const numNodes = treeSize(depth);
    const pool = new Int32Array(numNodes * 3);
    const half = Math.trunc(numNodes / 2);

    for (let i = numNodes - 1; i >= half; i--) {
        const base = i * 3;
        pool[base] = 1;
        pool[base + 1] = -1;
        pool[base + 2] = -1;
    }

    for (let i = half - 1; i >= 0; i--) {
        const base = i * 3;
        pool[base] = 0;
        pool[base + 1] = 2 * i + 1;
        pool[base + 2] = 2 * i + 2;
    }

    const stack = [0];
    let check = 0;

    while (stack.length > 0) {
        const nodeIdx = stack.pop();
        const base = nodeIdx * 3;
        check += pool[base];
        if (pool[base + 1] >= 0) {
            stack.push(pool[base + 1]);
            stack.push(pool[base + 2]);
        }
    }
    return check;
}

let total = buildAndCheck(MAX_DEPTH + 1);
const longLived = buildAndCheck(MAX_DEPTH);

for (let depth = 4; depth <= MAX_DEPTH; depth += 2) {
    let iterations = 1;
    for (let d = 0; d < MAX_DEPTH - depth + 4; d++) iterations *= 2;
    for (let i = 0; i < iterations; i++) {
        total += buildAndCheck(depth);
    }
}
total += longLived;
console.log(total);

File: bench/benchmarks/binary_trees/binary_trees.py
MAX_DEPTH = 14

def tree_size(depth):
    size = 1
    for _ in range(depth + 1):
        size *= 2
    return size - 1

def build_and_check(depth):
    num_nodes = tree_size(depth)
    pool = [0] * (num_nodes * 3)
    half = num_nodes // 2

    for i in range(num_nodes - 1, half - 1, -1):
        base = i * 3
        pool[base] = 1
        pool[base + 1] = -1
        pool[base + 2] = -1

    for i in range(half - 1, -1, -1):
        base = i * 3
        pool[base] = 0
        pool[base + 1] = 2 * i + 1
        pool[base + 2] = 2 * i + 2

    stack = [0]
    check = 0
    while stack:
        node_idx = stack.pop()
        base = node_idx * 3
        check += pool[base]
        if pool[base + 1] >= 0:
            stack.append(pool[base + 1])
            stack.append(pool[base + 2])
    return check

total = build_and_check(MAX_DEPTH + 1)
long_lived = build_and_check(MAX_DEPTH)

depth = 4
while depth <= MAX_DEPTH:
    iterations = 1
    for _ in range(MAX_DEPTH - depth + 4):
        iterations *= 2
    for _ in range(iterations):
        total += build_and_check(depth)
    depth += 2

total += long_lived
print(total)
